### PR TITLE
Try to retrieve missing auth events from multiple servers

### PR DESCRIPTION
**Pull request by neilalexander**
This tries to get missing auth events from multiple servers, and is a bit more tolerant of missing auth/state events on `/send`.

### CODE REVIEW

File: federationapi/routing/send.go [hunk 1]
```diff
@@ -367,53 +365,31 @@ func (t *txnReq) processEvent(ctx context.Context, e gomatrixserverlib.Event) er
 		return roomNotFoundError{e.RoomID()}
 	}
 
-	if len(stateResp.MissingAuthEventIDs) > 0 {
-		logger.Infof("Event refers to %d unknown auth_events", len(stateResp.MissingAuthEventIDs))
-
-		servers := []gomatrixserverlib.ServerName{t.Origin}
+	// We will need to know this when fetching missing auth or prev events.
+	getServers := func() {
+		t.servers = []gomatrixserverlib.ServerName{}
```
Comment (Kegsay) [inline federationapi/routing/send.go:370]: Feels like a peculiar form, why have this as a nested function which side effects by setting `t.servers`? Why not just have `t.getServers() []ServerName`?

File: federationapi/routing/send.go [hunk 2]
```diff
@@ -431,6 +407,49 @@ func (t *txnReq) processEvent(ctx context.Context, e gomatrixserverlib.Event) er
...
+				logger.WithError(err).Warnf("Failed to retrieve auth event %q", missingAuthEventID)
+				continue withNextServer
+			}
+			ev, err := gomatrixserverlib.NewEventFromUntrustedJSON(tx.PDUs[0], stateResp.RoomVersion)
+			if err != nil {
+				logger.WithError(err).Warnf("Failed to unmarshal auth event %q", missingAuthEventID)
+				continue withNextServer
+			}
+			if err = t.processEvent(ctx, ev); err != nil {
+				return fmt.Errorf("recursive t.processEvent: %w", err)
```
Comment (Kegsay) [inline federationapi/routing/send.go:440]: Do we really need to recurse? We don't care about state snapshots and events having prev_events for auth events right?

**Review by Kegsay — APPROVED** [review]

File: federationapi/routing/send.go (hunk 2 above)
Comment (neilalexander) [reply federationapi/routing/send.go:440]: Nah, I just hadn't changed it to `api.SendInputRoomEvents` yet. I'll do that.

File: federationapi/routing/send.go (hunk 1 above)
Comment (neilalexander) [reply federationapi/routing/send.go]: I've refactored it to be a bit less strange!

**Pull request merged**